src/rmt.rs

```diff
@@ -0,0 +1,415 @@
+//! Remote Control (RMT) module driver.
+//!
+//! The RMT (Remote Control) module driver can be used to send and receive infrared remote control
+//! signals. Due to flexibility of RMT module, the driver can also be used to generate or receive
+//! many other types of signals.
+//!
+//! This module is a wrapper around the [IDF RMT](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/peripherals/rmt.html)
+//! implementation. It is recommended to read before using this module.
+//!
+//! This is an initial implementation supporting:
+//!  * Transmission, currently only blocking.
+//!
+//! Not supported:
+//! * Receiving.
+//! * No channel locking. (To prevent users from accidentally reusing a channel.)
+//! * No buffer protection while transmitting.
+//! * Change of config after initialisation.
+//!
+//! # Example Usage
+//! ```
+//! use esp_idf_hal::gpio::Output;
+//! use esp_idf_hal::rmt::Channel::Channel0;
+//!
+//! let cfg = config:WriterConfig::new().clock_divider(1);
+//!
+//! let peripherals = Peripherals::take().unwrap();
+//! let led: Gpio18<Output> = peripherals.pins.gpio18.into_output().unwrap();
+//! let mut writer = Writer::new(led, cfg).unwrap();
+//!
+//! writer.push_pulse(Pulse::new(PinState::High, PulseTicks::max()));
+//! writer.push_pulse(Pulse::new(PinState::Low, PulseTicks::max()));
+//!
+//! writer.start().unwrap();
+//!
+//!```
+//!
+//! See the `examples/` folder of this repository for more.
+
+// TODO: Do we need to prevent users from creating two drivers on the same channel?
+// TODO: Should probably prevent writing to buffer while rmt is transmitting:
+//       "We must make sure the item data will not be damaged when the driver is still sending items in driver interrupt."
+use crate::gpio::OutputPin;
+use config::WriterConfig;
+use core::time::Duration;
+use embedded_hal::digital::PinState;
+use esp_idf_sys::{
+    esp, rmt_channel_t_RMT_CHANNEL_0, rmt_channel_t_RMT_CHANNEL_1, rmt_channel_t_RMT_CHANNEL_2,
+    rmt_channel_t_RMT_CHANNEL_3, rmt_config, rmt_config_t, rmt_config_t__bindgen_ty_1,
+    rmt_driver_install, rmt_get_counter_clock, rmt_item32_t, rmt_item32_t__bindgen_ty_1,
+    rmt_item32_t__bindgen_ty_1__bindgen_ty_1, rmt_mode_t_RMT_MODE_TX, rmt_tx_config_t,
+    rmt_write_items, EspError, EOVERFLOW, ESP_ERR_INVALID_ARG, RMT_CHANNEL_FLAGS_AWARE_DFS,
+};
+use std::convert::TryFrom;
+use std::mem::ManuallyDrop;
+
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+pub enum Channel {
+    // TODO: Work out the different number of channels per chip.
+    Channel0,
+    Channel1,
+    Channel2,
+    Channel3,
+}
+
+impl Into<u32> for Channel {
```

Comment:
[`From`, not `Into`](https://doc.rust-lang.org/std/convert/trait.Into.html).

Reply:
The aim here was to map a Channel into a `u32`, but this is going away for the singleton pattern.

Reply:
Np, I'm just saying that if you need "Into", you should implement "From". As the link I pasted in my original comment clarifies.

Reply:
Oh sorry, I see now!